File: sw/sim/eeprom.c

/*
 * Copyright 2021 Nicolas Maltais
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <sim/eeprom.h>
#include <sim/memory.h>
#include <sim/time.h>

#include <sys/eeprom.h>

#include <core/eeprom.h>
#include <core/trace.h>

#include <stddef.h>
#include <stdbool.h>

#define ERASE_BYTE 0xff

#ifdef SIM_MEMORY_ABSOLUTE
#define EEPROM_SIZE SYS_EEPROM_SIZE
#else
#define EEPROM_SIZE EEPROM_RESERVED_SPACE
#endif

#define STATUS_WREN_MASK 0x02
#define STATUS_BUSY_MASK 0x01

#define PAGE_MASK 0x1f  // page size = 32

enum {
    INSTRUCTION_WREN = 0x06,
    INSTRUCTION_WRDI = 0x04,
    INSTRUCTION_RDSR = 0x05,
    INSTRUCTION_WRSR = 0x01,
    INSTRUCTION_READ = 0x03,
    INSTRUCTION_WRITE = 0x02,
};

sim_mem_t* eeprom;

static struct {
    uint8_t status;
    uint8_t instr;
    eeprom_t address;
    size_t pos;
    bool writing;
} spi_eeprom;


void sim_eeprom_init(void) {
    eeprom = sim_mem_init(EEPROM_SIZE, ERASE_BYTE);
    // in simulation, EEPROM always starts at address 0, but has the correct size.
    sys_eeprom_set_location(0, EEPROM_SIZE);
}

void sim_eeprom_free(void) {
    sim_mem_free(eeprom);
    eeprom = 0;
}

void sim_eeprom_load(const char* filename) {
    sim_mem_load(eeprom, filename);
}

void sim_eeprom_save(void) {
    sim_mem_save(eeprom);
}

void sim_eeprom_spi_transceive(size_t length, uint8_t data[static length]) {
    if (!eeprom) {
        return;
    }

    for (size_t i = 0; i < length; ++i) {
        uint8_t b = data[i];
        size_t pos = spi_eeprom.pos;

        if (spi_eeprom.instr == 0 && pos == 0) {
            // set new instruction
            spi_eeprom.instr = b;
        }

        switch (spi_eeprom.instr) {
            case INSTRUCTION_WREN: {
                spi_eeprom.status |= STATUS_WREN_MASK;
                break;
            }
            case INSTRUCTION_WRDI: {
                spi_eeprom.status &= ~STATUS_WREN_MASK;
                break;
            }
            case INSTRUCTION_RDSR: {
                if (pos == 1) {
                    data[i] = spi_eeprom.status;
                    // after write, status will read as busy once, then ready.
                    spi_eeprom.status &= ~STATUS_BUSY_MASK;
                }
                break;
            }
            case INSTRUCTION_WRSR: {
                spi_eeprom.status = b & 0x8c;
                break;
            }
            case INSTRUCTION_READ: {
                if (pos == 1) {
                    spi_eeprom.address = b << 8;
                } else if (pos == 2) {
                    spi_eeprom.address |= b;
                    spi_eeprom.address %= SYS_EEPROM_SIZE;
                } else if (pos >= 3) {
                    data[i] = eeprom->data[spi_eeprom.address];
                    spi_eeprom.address = (spi_eeprom.address + 1) % SYS_EEPROM_SIZE;
                }
                break;
            }
            case INSTRUCTION_WRITE: {
                if (pos == 0) {
                    spi_eeprom.writing = (spi_eeprom.status & STATUS_WREN_MASK) != 0;
                    spi_eeprom.status &= ~STATUS_WREN_MASK;
                    spi_eeprom.status |= STATUS_BUSY_MASK;
                } else if (pos == 1) {
                    spi_eeprom.address = b << 8;
                } else if (pos == 2) {
                    spi_eeprom.address |= b;
                    spi_eeprom.address %= SYS_EEPROM_SIZE;
                } else if (spi_eeprom.writing) {
                    eeprom->data[spi_eeprom.address] = b;
                    spi_eeprom.address = (spi_eeprom.address & ~PAGE_MASK) |
                                         ((spi_eeprom.address + 1) & PAGE_MASK);
                }
                break;
            }
            default:
                if (pos == 0) {
                    trace("unsupported EEPROM instruction 0x%02x", spi_eeprom.instr);
                }
                break;
        }

        ++spi_eeprom.pos;
    }
}

void sim_eeprom_spi_reset(void) {
    spi_eeprom.instr = 0;
    spi_eeprom.address = 0;
    spi_eeprom.pos = 0;
    spi_eeprom.writing = false;
}

File: sw/utils/prog/eeprom.py
#  Copyright 2022 Nicolas Maltais
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

# This module is used to read, write & erase the EEPROM chip on the game console
# The EEPROM chip model number is FT25C32A
# Datasheet: https://www.fremontmicro.com/downfile.aspx?filepath=/upload/2019/0805/1604ntvhj2.pdf&filename=ft25c32a_ds_rev0.82.pdf

from pathlib import Path
from typing import Union

from prog.memory import MemoryDriver, MemoryLocal
from prog.spi import SpiInterface, SpiPeripheral
from utils import ProgressCallback

EEPROM_SIZE = 4096
# eeprom page size in bytes
PAGE_SIZE = 32

ERASE_BYTE = 0xff

# instruction set
INSTR_WREN = 0x6
INSTR_WRDI = 0x4
INSTR_RDSR = 0x5
INSTR_WRSR = 0x1
INSTR_READ = 0x3
INSTR_WRITE = 0x2


class EepromDriver(MemoryDriver):
    """Class used to read & write to the EEPROM chip."""
    spi: SpiInterface

    def __init__(self, spi: SpiInterface):
        spi.peripheral = SpiPeripheral.EEPROM
        self.spi = spi

    def get_size(self) -> int:
        return EEPROM_SIZE

    def requires_block_erase(self) -> bool:
        return False

    def read(self, address: int, count: int, progress: ProgressCallback) -> bytes:
        self._wait_ready()
        command = [INSTR_READ, address >> 8, address & 0xff]
        # insert dummy MOSI bytes
        for i in range(count):
            command.append(0)
        received = self.spi.transceive(command, progress=lambda c, t: progress(c - 3, t - 3))
        return received[3:]

    def write(self, address: int, data: bytes, progress: ProgressCallback) -> None:
        """Write bytes to EEPROM at an address."""
        # wait until ready and enable write latch
        written = 0
        it = iter(data)
        while written < len(data):
            self._wait_ready()
            self.spi.transceive([INSTR_WREN])
            # write data for one page
            command = [INSTR_WRITE, address >> 8, address & 0xff]
            page_end = (address & ~(PAGE_SIZE - 1)) + PAGE_SIZE
            while written < len(data) and address < page_end:
                byte = next(it, None)
                if byte is None:
                    break
                command.append(byte)
                written += 1
                address += 1
            self.spi.transceive(command)
            progress(written, len(data))

    def erase(self, progress: ProgressCallback) -> None:
        data = bytearray()
        data[:] = ERASE_BYTE
        self.write(0, data, progress)

    def _wait_ready(self) -> None:
        """Wait until status register indicates ready status"""
        while True:
            received = self.spi.transceive([INSTR_RDSR, 0])
            if not (received[1] & 0x1):
                return

    @staticmethod
    def local(filename: Union[str, Path]) -> MemoryLocal:
        return MemoryLocal(filename, EEPROM_SIZE, ERASE_BYTE, 1, False)
